Refactor models to be entities

rmainwork opened the pull request:
The purpose of this PR is to refactor the user entity currently in the repo (introduced in #5) to be in a format able to be used with TypeORM. The user entity as things currently stand expects it's data to be provided via constructor argument which causes problems with TypeORM. This is because TypeORM doesn't pass it's values through the constructor, which causes exceptions to be raised when entities are instantiated. Since TypeORM also allows for the ActiveRecord pattern where properties are set on entities and then saved, this may make the class incompatible with TypeORM.

Adding to that, the usage of getters on entity objects causes problems when instanciating them as TypeScript seems to expect the getter to be set at call time rather than dynamically by the values of the other fields it exposes. Calculated fields do exist in TypeORM, however I haven't included them in this PR as the current repo state does not include TypeORM and it felt a little out of scope.

## Types of changes
<!--
  What types of changes does your code introduce? Put an `x` in all the boxes
  that apply:
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality
      to change)

## Checklist:
<!--
  Go over all the following points, and put an `x` in all the boxes that apply.
-->
- [x] I have run `eslint` on the code
- [x] I have added JSDoc for all of my code (where applicable)
- [x] I have added tests to cover my changes.

## Priority:
- [X] Normal <!-- New piece of functionality -->
- [ ] High <!-- Critical bug requiring urgent review -->

## Related Issues:
<!--
  Add the issue number in below that this PR is intended to address.

  Also mention any issues that this PR is related to, and if possible,
  provide more information regarding the relationship to the issues in
  question (i.e does the PR fix the issue, is it designed to fix another bug
  that is similar to the issue etc.)
-->
N/A

<!--
  Attribution:
  PR Template adapted from: https://github.com/h5bp/html5-boilerplate/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

Commit:
Move user entity class to entities folder

Commit:
Refactor user class to be a TypeORM compatible entity

jonseitz commented on src/client/components/App.tsx:
```diff
@@ -53,9 +53,9 @@ const ColdApp: SFC = (): ReactElement => {
         return user;
       })
-      .then((user): void => {
+      .then(({ firstName, lastName }): void => {
```
This `user` value should be of type `UserReponse`, which, as defined in `src/common/dto/users/userResponse.dto.ts`, does have both `fullName` and `listName` properties. If those will not be present the dto should be changed, or if the dto is not changing this should be reverted.

rmainwork replied:
Fixed in 4a9ab5092e095052117b86233c7345fb1b56d781

jonseitz commented on src/client/components/__tests__/App.test.tsx:
```diff
@@ -33,8 +33,10 @@ describe('App', function () {
-        return waitForElement(() => getByText(fullName, { exact: false }));
+        const { firstName, lastName } = dummy.regularUser;
+        return waitForElement(
+          () => getByText(`${firstName} ${lastName}`, { exact: false })
```
see above re: the `UserResponse` dto

rmainwork replied:
Fixed in 4a9ab5092e095052117b86233c7345fb1b56d781

jonseitz commented on src/common/__tests__/data/users.ts:
```diff
@@ -1,10 +1,10 @@
 };
 
-export const regularUser: User = new User(rawRegularUserData);
+export const regularUser = rawRegularUserData;
```
This separation of `rawRegularUserData` and `regularUser` is only necessary when there's a separate `User` class. If there isn't one, there should just be one export of `regularUser: User`

rmainseas replied:
Fixed in 2560bd5a21a3fb05b10af708a4173494bbf11607

rmainwork replied:
Fixed in 2560bd5a21a3fb05b10af708a4173494bbf11607

jonseitz commented on src/common/__tests__/data/users.ts:
```diff
@@ -1,10 +1,10 @@
   firstName: 'Regular',
   lastName: 'User',
-  id: '88888888',
+  id: '84ac7431-f612-49c4-953c-0497f44cebc2',
```
This `id` field will actually be the `eppn` returned by Harvard Key, not the `uuid` since we won't be storing users in the database

rmainwork replied:
Refactored in 7bd2ec93d62c13803692fb007615f6925b53081d

jonseitz commented on src/server/services/saml.strategy.ts:
```diff
@@ -24,22 +24,24 @@ class SAMLStrategy extends PassportStrategy(Strategy) {
       throw new UnauthorizedException('You are not authorized to use this application. Please contact SEAS computing');
     }
-    return new User({
+    const authenticatedUser: User = {
```
So what's the difference between declaring these as `User`s with types vs. instantiating them with a constructor? Is that just necessary for TypeORM, or is there some other benefit?

rmainwork replied:
It's for TypeORM, yeah. TypeORM (I think) instantiates them internally by doing `new Foo()` and then setting properties. So when it does that, no values are passed to the constructor, causing the constructor to throw.

jonseitz reviewed (changes requested):
I have to be honest, I'm a little unclear on what this PR is actually addressing (it's marked as a bugfix but there's no associated issue). I think it's good practice to provide quick recap of changes and an explanation of why with every PR, both for the current reviewers sake as well as providing a record for the future.

Also, my understanding was that we wouldn't be maintaining our own database of user, and instead we'd be relying fully on Harvard Key for Authentication and Grouper for Authorization. In that case, would our user entity need to be compatible with TypeORM?

rmainwork commented:
> I have to be honest, I'm a little unclear on what this PR is actually addressing (it's marked as a bugfix but there's no associated issue).

Yeah - that's an oversight on my part I think. It's intended to address the slight discrepancy between the way TypeORM expects things to be and the way things actually are in our repo. I don't want to specialize too much on TypeORM in the boilerplate, since the boilerplate doesn't include TypeORM....but I also don't want to make it a headache to integrate later.

> I think it's good practice to provide quick recap of changes and an explanation of why with every PR, both for the current reviewers sake as well as providing a record for the future.

That makes sense, I've added that above.

> Also, my understanding was that we wouldn't be maintaining our own database of user, and instead we'd be relying fully on Harvard Key for Authentication and Grouper for Authorization. In that case, would our user entity need to be compatible with TypeORM?

I believe this to be the case, however the `User` entity was introduced in #5. That being the case, it makes sense for it to exist in a format that is compatible with the technologies we're likely to be using going forward. Plus, one could argue that it sits better in an `entities` folder than a `models` folder since it doesn't really perform any kind of database communication, it's just a static representation of a type. We may not necessarily have a table for it(although the current ERD for course planning says that we do in order to relate "views" to authenticated users), but it probably makes sense to at least have the shape of a user in there, even if we never build out an associated database model and/or migration to go with it.

Commit:
Update user DTO to match user entity

Commit:
Merge `rawRegularUserData` and `regularUser`

The former was only used as a way of separating out the user data from the user object instantiation as is no longer required

Commit:
Replace `id` with `eppn`

Presence  of an `id` field on `User` has the potential to be misleading and suggest that it represents a database PK id. Since users are not stored in the database this is not the case. Instead, `eppn` is a much better name for the field and more accurately describes it's purpose

Commit:
Add doc blocks for user entity and harvard key interface

jonseitz commented:
Let's discuss this when we're both in the office on Friday. Overall I agree with the changes, but I'd like to talk through this a little more and make sure I understand this approach before we commit to it.

rmainwork commented:
I need to go back and refactor the `User` entity to re-introduce getters. Also need to accept optional `data` param through constructor.

Since we won't be using this with TypeORM, it shouldn't cause any compatibility issues.

Commit:
Replace user constructor and accesor properties

These were originally removed as they were incompatible with TypeORM, but since this object will not be used with TypeORM - it is fine to re-add them.

Commit:
Test constructor sets all fields to empty string by default

jonseitz reviewed (approved):
This is ready to be merged.
